### ISSUE TITLE
mas outdated not working

### ISSUE THREAD

**Issue by misc-activities** [mas-cli/mas#306]
I just installed mas on my macOS Catalina installation. I have signed into the app store from the Mac App Store. When I run `mas list`, I can see all the installed apps. There is an update available for Xcode which I want to try installing using mas. However, when I run `mas outdated`, it doesn't show the available update. I ran `mas reset` and then ran `mas outdated` again but that didn't work as well.

**Comment by hernancorigliano** [comment]
I have the same issue but not only with Xcode, but all pending updates. There is no output from `mas outdated`.

**Comment by Luminus** [comment]
I'm experiencing the same situation.

* `mas outdated` has no output even though I have apps that need to be updated
* Similarly, `mas upgrade` returns `Everything is up-to-date`, which is not true
* Trying to update individual apps by running `mas upgrade XXXXXXXX` where XXXXXXXX is the relevant ID returns `Warning: Nothing found to upgrade`

**Comment by andrewcrook** [comment]
Doesn’t seem to work on Catalina I have never got it to work fully.
I can only get list, install and remove to work.
At least I can install all my applications scripted on Big Sur
the question is do all the Mas features work on Big Sur or is it going to be like Catalina?

**Comment by Yelles** [comment]
same for me, updates are available in AppStore, but when I run mas outdated it doesn't show the available update... same for mas upgrade/install
(i'm on catalina)

**Comment by drewauff** [comment]
Outdated switch dead on Catalina.

**Comment by ke11** [comment]
not working on Catalina after the recent security patch

**Comment by brian163** [comment]
I may be seeing something similar on Mojave. I came here to see if I could figure out why the App Store app is telling me there is an update for BlackMagic Speed Disk (425264550) but mas (which I usually run to avoid the aggravation with the GUI I've had in the past) just told me everything was up to date.

When I do a manual list of everything installed:
mas list | grep Disk
425264550 Disk Speed Test (3.2)

When I tell mas to search the store:
mas search "Disk Speed Test" | grep Disk 
425264550  Blackmagic Disk Speed Test              (3.3)

When I just run mas outdated, nothing is returned. And mas upgrade does nothing.

Obviously there is a name change there but I wouldn't think that would cause mas to mismatch since the app ID are the same?

**Comment by Stanzilla** [comment]
Also not working on BigSur

**Comment by andrewcrook** [comment]
Apple has changed the way the AppStore works this may mean mas might not ever work with it.

**Comment by benftrex** [comment]
Not working for me on Catalina and Bigsur. 
As others have pointed, only "mas list" works. Outdated and upgrade don't, and never have for me since Catalina.

**Comment by lonelykid** [comment]
> I have the same issue but not only with Xcode, but all pending updates. There is no output from `mas outdated`.
> 
> Looks like `CKUpdateController.shared().availableUpdates()` is always returning an empty list.

If ``mas info`` and ``mas list`` works, isn't it an easy fix to use the info to decide if an upgrade is needed? This might be clumsy and slower but at least it works.

**Comment by rgoldberg** [comment]
> If mas info and mas list works, isn't it an easy fix to use the info to decide if an upgrade is needed

I just verified that `mas info` shows the current version of the app in the store while `mas list` shows the version of the app installed on my machine, so this should work.

I'd be willing to work on fixing this, but I haven't used Xcode, Swift, or Mac APIs before (I mainly code in Java & Kotlin).

I forked the repo, then received the following 8 errors & 1 warning when building the project in Xcode 12.3 (12C33) on macOS 11.1 on Intel.  If someone can help me fix the build problems, I'll continue working on a PR for this issue.

```
~/Code/fork/mas/mas/main.swift:9:8:
No such module 'Commandant'

~/Code/fork/mas/MasKit/Commands/Account.swift:9:8:
No such module 'Commandant'

SwiftLint not installed, download from https://github.com/realm/SwiftLint

~/Code/fork/mas/Carthage/Build/Mac/Commandant.framework:
No such file or directory

~/Library/Developer/Xcode/DerivedData/mas-gwgogfusotbawbfpkrdgrxttzbny/Build/Intermediates.noindex/mas-cli.build/Debug/MasKit.build/module.modulemap:9:12:
Header 'MasKit-Swift.h' not found

~/Code/fork/mas/MasKitTests/Commands/AccountCommandSpec.swift:9:18: Could not build Objective-C module 'MasKit'

~/Code/fork/mas/Carthage/Build/Mac/Quick.framework:
No such file or directory

~/Code/fork/mas/Carthage/Build/Mac/Nimble.framework:
No such file or directory

~/Code/fork/mas/Carthage/Build/Mac/Commandant.framework:
No such file or directory
```

**Comment by lonelykid** [comment]
> > If mas info and mas list works, isn't it an easy fix to use the info to decide if an upgrade is needed
> 
> I just verified that `mas info` shows the current version of the app in the store while `mas list` shows the version of the app installed on my machine, so this should work.
> 
> I'd be willing to work on fixing this, but I haven't used Xcode, Mac APIs, Swift, or Objective-C before (I mainly code in Java & Kotlin).
> 
> I forked the repo, cloned the fork, and opened the project in Xcode 12.3 (12C33) on macOS 11.1 on Intel.
> 
> I received the following 8 errors & 1 warning when I built.
> 
> If someone can help me fix the build problems, I'll continue working on a PR for this issue.
> 
> ```
> ~/Code/fork/mas/mas/main.swift:9:8:
> No such module 'Commandant'
> 
> ~/Code/fork/mas/MasKit/Commands/Account.swift:9:8:
> No such module 'Commandant'
> 
> SwiftLint not installed, download from https://github.com/realm/SwiftLint
> 
> ~/Code/fork/mas/Carthage/Build/Mac/Commandant.framework:
> No such file or directory
> 
> ~/Library/Developer/Xcode/DerivedData/mas-gwgogfusotbawbfpkrdgrxttzbny/Build/Intermediates.noindex/mas-cli.build/Debug/MasKit.build/module.modulemap:9:12:
> Header 'MasKit-Swift.h' not found
> 
> ~/Code/fork/mas/MasKitTests/Commands/AccountCommandSpec.swift:9:18:
> Could not build Objective-C module 'MasKit'
> 
> ~/Code/fork/mas/Carthage/Build/Mac/Quick.framework:
> No such file or directory
> 
> ~/Code/fork/mas/Carthage/Build/Mac/Nimble.framework:
> No such file or directory
> 
> ~/Code/fork/mas/Carthage/Build/Mac/Commandant.framework:
> No such file or directory
> ```

Looks like you are just missing some package, for example, https://github.com/Carthage/Commandant

**Comment by rgoldberg** [comment]
A git repo for Commandant already exists under `./Carthage/Checkouts/Commandant` in the `mas` repo.

So the files should be there, it's just that the build doesn't know where they are.


Are there build setup instructions for `mas`?

I didn't see any in the readme or contributing docs.

I'm used to the build system (Gradle, for my projects) automatically downloading listed dependencies and producing the correct classpath for the compiler.

Do Apple ecosystem builds really require manual setup these days?

**Comment by lonelykid** [comment]
> A git repo for Commandant already exists under `./Carthage/Checkouts/Commandant` in the `mas` repo.
> 
> So the files should be there, it's just that the build doesn't know where they are.
> 
> Are there build setup instructions for `mas`?
> 
> I didn't see any in the readme or contributing docs.
> 
> I'm used to the build system (Gradle, for my projects) automatically downloading listed dependencies and producing the correct classpath for the compiler.
> 
> Do Apple ecosystem builds really require manual setup these days?

Try the build script maybe. https://github.com/mas-cli/mas/blob/master/script/build

**Comment by andrewcrook** [comment]
Looks like you need Ruby with Bundler, Brew and Xcode 

The, to me,  it looks like you to do the following first:

- use the brewfile - `brew Brewfile install`
- make init uses `swiftenv `to install the correct version of swift.
- then `make bootstrap` to install all the dependencies 

Looks like you need to run  Brew on brewfile to install dependences  this includes mint a swift dependency manager 
The mint file includes Carthage a cocoa dependency manager the Carfile includes the Commandant command line framework.

ruby packages are also managed Bundler with a gem file

Brewfile also installs make 

so looking at the Makefile it looks like the build commands are initiated through the make tool
and make runs the appropreate scripts in the script directory. 

The bootstrap option seems to call all the dependency managers in turn to setup the environment in which to build.

**Comment by rgoldberg** [comment]
Is the Apple ecosystem so poor that Xcode cannot recognize standard build scripts & run without manual configuration?

Or is this project just an oddity?

If Xcode does recognize standard build scripts, but many projects don't use them, then that's still a failure of the Apple ecosystem.

**Comment by andrewcrook** [comment]
Xcode has every thing to build products for Apple devices it uses project files and a majority of app developers just use Xcode.

There are also many developers who prefer using other tools with the Xcode command line tools.
Developers may like like to build custom dev/ environments , automated a host of other tasks, different versioning, different code management , target different platforms/architectures,  customise their build CI/CD pipelines and use third party services. 

Swift is also open sourced so some developers are using it server side with a host of different open source tools.

One thing to consider is that MAS might be using non-public apis.

**Comment by rgoldberg** [comment]
I have a basic version of `mas outdated` working in Xcode debug.  It isn't working yet on the command line (I assume because of dylib selection issues), but I'll deal with this later (I haven't previously had to deal with dylib issues since I normally develop for the jvm, and I haven't ever used Xcode before).

A few questions: is different local & store version strings enough to consider an app outdated? Or must I implement a version ordering scheme to ensure that the store version is newer, not just different?

If no store app is found for the identifier for a local app, should I output an error to stderr?  Should I silently ignore?  Should I do something else?  Should I output red (or another color) text for errors, or just output plain text?  Should errors be output as they occur, or should they be saved to all be output after non-error output?

There might be some rough edges.  I'm only building for x86_64 (instead of for both x86_64 & arm64) on my Intel Mac due to some error(s).  I don't know if existing mas won't build for arm64, or if something is wrong with my Xcode setup.  etc.

The sooner I have answers to the above, the sooner I'll be able to submit a PR (which will probably have to be revised, sue to my inexperience with all the associated tech).

**Comment by lonelykid** [comment]
I think you need to mimic availableUpdates as much as you can. Not sure how it behaves since it's undocumented. Basically what outdated produces should be the same as what AppStore prompt you to update. I never encountered situations that app version is reverted in appstore. But it seems no harm to add. For removed app, a message would be good to have.

**Comment by rgoldberg** [comment]
Are versions guaranteed to only contain digits & dots? If not, ordering can get tricky. e.g., the maven version ordering algorithm is massively broken. I don't want to create an ordering that differs from what the store / publishers use.

I would want to know about all version discrepancies between local & store, even if some would be considered reversions.

**Comment by rgoldberg** [comment]
I found the `printWarning` & `printError` functions in `Utilities.swift`.

They seem to output to stdout.

It seems suboptimal for `mas outdated` to output warnings about not finding an app in the store to stdout along with normal output that indicates that an app has an available update.

But, it would make sense to use the same formatting for warnings & errors for this command.

Should all warnings & errors be output to stderr?  Should only the warnings from my updated `outdated` command go to stderr?  Or should I just send the warnings to stdout?

I'll use a warning for an app not being found, and let all other error conditions be handled by the existing error handling setup.

As an aside, there seems to be a typo in `printError`: if not using colored output, it outputs "Warning: " instead of "Error: ".  Must I create a separate PR to fix that, or can I include it in this PR?

**Comment by rgoldberg** [comment]
I created a PR that fixes both `outdated` & `upgrade`.

It might be rough around the edges.

Please review & provide feedback.

I'll improve it as directed.

**Comment by rgoldberg** [comment]
Is there anything that I must do to get my PR merged?

**Comment by lonelykid** [comment]
It seems the repo owner is just not around. Contact @phatblat directly maybe

**Comment by rgoldberg** [comment]
I hope they're Ok.

**Comment by rgoldberg** [comment]
@lonelykid Can any users besides @phatblat merge pull requests?

Does anyone know if @phatblat is OK? He seems to be MIA…

**Comment by lonelykid** [comment]
> @lonelykid Can any users besides @phatblat merge pull requests?
> 
> Does anyone know if @phatblat is OK? He seems to be MIA…

I don't think other people has write access. He is certainly around. Just check his commit record on his github profile and twitter. So it's good that he is OK, maybe just lost interest in maintaining this project. Can you try send him a message on twitter?

**Comment by rgoldberg** [comment]
I would assume that he gets notifications about mentions in github, so I don't want to rope in his Twitter if he's actively ignoring this project.

I guess he's just very busy.

**Comment by andrewcrook** [comment]
> Does anyone know if @phatblat is OK? He seems to be MIA…

@lonelykid  I read in another thread that @phatblat had to have surgery on his wrist(s) and is only just getting back into typing code again.  He has just released 1.8.0 which works on ARM64 (it used to just segfault), however, I am still having this original issue from way back as Intel Macs with Catalina (I now have Apple Silicon and Big Sur)  with the outdated and upgrade not working as expected. Now he is back perhaps the ball can start to roll again. PS thank you for all your time you have put into this.

**Comment by phatblat** [comment]
Should be fixed now in [1.8.1](https://github.com/mas-cli/mas/releases/tag/v1.8.1). @misc-activities can you confirm?

**Comment by lonelykid** [comment]
@rgoldberg Hmmm. Seems like mas info sometimes does not give the newest version. Could you check your result of `mas info 1249901692`. 

Another issue, if you run  the same `mas info xx` twice. The second time will incur an error but still produces the same output. This is not related but could you check if you have the same problem?
``
2021-03-14 23:37:53.169 mas[53764:1108974] INFO: fetch-response is unable to open the file /Users/username/Library/Caches/com.mphys.mas-cli/fsCachedData/CF392AD7-531A-4927-858F-0344761770C8. Errno: 2
``

**Comment by misc-activities** [comment]
I tested by installing from the .pkg file. I had to run `mas outdated` twice to see the updates. When I ran it the first time, it didn't show any updates. All available updates were shown correctly the second time I ran the command.

**Comment by rgoldberg** [comment]
@lonelykid Confirmed that `mas info 1249901692` seems to list an out-of-date version, as does `mas search ScanScan`, and possibly other commands.  I didn't change those commands, so that is an existing issue that should be created.

`mas info xx` twice didn't give me the error.  But `mas info` on specific IDs consistently gives me the error, like `mas info 409203825`.

The file doesn't exist, but one with the same name but with a `.tmp` extension appended does exist. I traced the issue to the line `_ = semaphore.wait(timeout: .distantFuture)` in `loadDataSync` in `NetworkManager.swift`.  This is an existing issue that should also be created.

**Comment by rgoldberg** [comment]
@misc-activities I don't have any outdated apps, so I cannot currently confirm.

Must you run it twice on every new terminal?  Every login?  Every reboot? Or just one extra time once, and all subsequent runs work across new terminals, logins, reboots, etc.?

**Comment by Luminus** [comment]
I've just updated to 1.8.1 and everything seems to work fine for me now.

- `mas outdated` listed my outdated apps
- `mas upgrade` upgraded all of them
- I ran `mas info` on the same app 3 times with no errors
- `mas info 1249901692` does list ScanScan version 3.1.1 instead of 3.1.3 that's shown in the App Store app. I'm not sure what the deal is with that, but I've never used this app would have never tested this if @lonelykid hadn't mentioned it above.

Thanks for sorting this out @phatblat

**Comment by lonelykid** [comment]
what`mas info 1249901692` listed in a not only outdated, but also deleted version. I remember something similar happened in the discussion of PR 46ebd7c9e4efa6a37b806986764b60a57f0b568c. It seems mas info will have trouble when the author revoked a version and then published newer version.

**Comment by Luminus** [comment]
@lonelykid `mas info 409203825` runs without errors for me. I've just run it 6 times now to test and I can't report any issues.

`mas info 1462114288` returns `Grammarly for Safari 9.29` for me.

Are you running Big Sur? I'm wondering if there could be an OS version issue compounding things on your end.

The situation with `mas info 1249901692` is still a complete mystery to me, though.

**Comment by lonelykid** [comment]
@Luminus I'm using Big Sur 11.2.3. What's your os version?

**Comment by Luminus** [comment]
@lonelykid Big Sur 11.2.2. I'm about to run the update to 11.2.3 now.

**Comment by misc-activities** [comment]
@rgoldberg It's just one extra time once.

**Comment by lonelykid** [comment]
@Luminus That's strange. I don't think it's bug introduced by new macos version. but good luck.

@rgoldberg Could you try `mas info 422293948`? This never give me the error because somehow mas does not write cache file at all for this app. This seems to be a cache issue. After one run, the cache file is added `.tmp` extension. the next time mas will report error and write correct cache file. and the subsequent time will run without issue.

**Comment by rgoldberg** [comment]
I'll try everything people need me to try later today.

I think we should probably open 2 (or more) new issues and have the discussions in the relevant issue, however, instead of continuing here, since they seem to be existing issues with common code that was originally used by `info` & `search` (& possibly other commands). `outdated` & `upgrade` now use that same common code instead of relying on a different long-broken Apple API.

**Comment by bevanjkay** [comment]
I'm not sure if this is relevant to MAS or an App Store issue, but I have an up to date app showing for the mas outdated command with a version number formatting difference.

```
mas outdated 
571213070 DaVinci Resolve (17.1.0 -> 17.1)
```

I have tried updating through mas but it is still showing up as outdated.

```
mas info 571213070
DaVinci Resolve 17.1 [0.0]
By: Blackmagic Design Inc
Released: 2021-03-10
Minimum OS: 10.14.6
Size: 1.83 GB
From: https://apps.apple.com/us/app/davinci-resolve/id571213070?mt=12&uo=4
```

The installed app shows as version 17.1.0 in the 'get info' window.

**Comment by ondrejfuhrer** [comment]
I have a similar issue to one posted by @Luminus 

```
❯ mas outdated 
1529448980 Reeder (5.0.1 -> 5.0.6)
❯ mas info 1529448980
Reeder 5. 5.0.6 [9.99]
By: Silvio Rizzi
Released: 2021-03-17
Minimum OS: 10.15
Size: 10,1 MB
From: https://apps.apple.com/us/app/reeder-5/id1529448980?mt=12&uo=4
```

Even though I'm running the updated version, `was outdated` still states the old version 🙁

**Comment by lonelykid** [comment]
> I have a similar issue to one posted by @Luminus
> 
> ```
> ❯ mas outdated 
> 1529448980 Reeder (5.0.1 -> 5.0.6)
> ❯ mas info 1529448980
> Reeder 5. 5.0.6 [9.99]
> By: Silvio Rizzi
> Released: 2021-03-17
> Minimum OS: 10.15
> Size: 10,1 MB
> From: https://apps.apple.com/us/app/reeder-5/id1529448980?mt=12&uo=4
> ```
> 
> Even though I'm running the updated version, `was outdated` still states the old version 🙁

This seems a different issue. Yours didn't show incorrect installed version, but others' show incorrect new version. Did you upgrade the app outside AppStore? If so, try reinstalling it.

**Comment by ondrejfuhrer** [comment]
> This seems a different issue. Yours didn't show incorrect installed version, but others' show incorrect new version. Did you upgrade the app outside AppStore? If so, try reinstalling it.

Ah, I haven't realised that. I upgraded the app through `mas upgrade`. Will that make it inconsistent?

**Comment by andrewcrook** [comment]
> This seems a different issue. Yours didn't show incorrect installed version, but others' show incorrect new version. Did you upgrade the app outside AppStore? If so, try reinstalling it.

I believe there is a way to tell if an application came from the App Store or not via the apps code signature
mas should all have this feature? if it doesn't keep checking that an app is from the App Store it should do because it is so easy to install a version outside the store over the same app from the App Store and vice versa.

**Comment by ondrejfuhrer** [comment]
> I believe there is a way to tell if an application came from the App Store or not via the apps code signature.

Ah, maybe I also misunderstood the question asked earlier, so just to clarify this as well - that app I installed and always updated though AppStore before. Just last time I upgraded it I used / tested `mas upgrade` for it.

**Comment by captcarl13** [comment]
Also having similar issue after running `mas outdated`. macOS 11.2.3 (20D91):

`Warning: Identifier 490152466 not found in store. Was expected to identify iBooks Author.
Warning: Identifier 1175706108 not found in store. Was expected to identify Geekbench 4.
897118787 Shazam (2.10.0 -> 2.10)
1509611686 Spaces (11.4 -> 8.3)
Warning: Identifier 1148868081 not found in store. Was expected to identify iStat View.
405399194 Kindle (1.30.0 -> 1.30)
1171820258 Highland 2 (2.9.7 -> 2.9.8)
973039644 The Last Campfire (1.6 -> 1.4)
1466040723 Creaks (1.0.5 -> 1.0.4)`

`mas upgrade` nets cache errors. macOS caching service is enabled on target computer:
`2021-03-19 07:49:28.474 mas[95339:8645511] INFO: fetch-response is unable to open the file /Users/capt_carl/Library/Caches/com.mphys.mas-cli/fsCachedData/BD378A59-7682-4969-A91A-5E5A619E469D. Errno: 2
Upgrading 5 outdated applications:
Shazam (2.10.0), Spaces (11.4), Kindle (1.30.0), The Last Campfire (1.6), Creaks (1.0.5)`

**Comment by chris-araman** [comment]
The `fetch-response` issue mentioned above and also reported as #328 and #337 should be fixed in #345 and is expected to be included in the next release.

**Comment by chris-araman** [comment]
Actually, I think my fix to stop caching search results might have resolved this issue around `outdated`.

```
caraman@mac ~/s/mas (master)> mas outdated
1480068668 Messenger (97.11.116 -> 96.0)
2021-04-16 13:57:14.492 mas[69370:1333221] INFO: fetch-response is unable to open the file /Users/caraman/Library/Caches/com.mphys.mas-cli/fsCachedData/28A4F771-D9CF-4CA0-91CB-43367904B6E6. Errno: 2
caraman@mac ~/s/mas (master)> /Users/caraman/Library/Developer/Xcode/DerivedData/mas-ghhkgvjzsmagpaffvtwdwjazwfnt/Build/Products/Debug/mas outdated
1480068668 Messenger (97.11.116 -> 96.0)
1295203466 Microsoft Remote Desktop (10.6.0 -> 10.6.1)
```

Obviously, the mismatch between Messenger's bundle version and App Store version is still a problem (and is tracked in #331). However, the Remote Desktop update is correctly reported by the version of `mas` with #345. @phatblat, I think it might be worth releasing this as 1.8.2 even if some additional issues remain.

**Comment by chris-araman** [comment]
The remaining discrepancies from `outdated` should be resolved by #346. If folks who were seeing these issues could try the upcoming release and report back any new issues, I'd appreciate it.
